In [ ]:
book1 = 'my book'
book2 = 'your book'
books = ['all_books']
books.append(book1)
books.append(book2)
#print(books)
books.clear()
#print(books)

my_books = ['book1', 'book2', 'book3', 'book4', 'book5']
tom_books = my_books.copy()
#print(my_books, tom_books)

In [ ]:
my_orders = ['order1', 'order2', 'order3', 'order4', 'order5']
anne_orders = my_orders[:]
print(my_orders, anne_orders)

['order1', 'order2', 'order3', 'order4', 'order5'] ['order1', 'order2', 'order3', 'order4', 'order5']


In [ ]:
random_values = [3, 5, 0, -1, 2, 10, 15, -5]
random_values.sort()
print(random_values)

[-5, -1, 0, 2, 3, 5, 10, 15]


In [ ]:
place_and_money = {1: 100, 2: 50, 3: 10}
a = place_and_money.pop(3)
print(a)

10


In [ ]:
test_dich3 = {}
test_dich3.update({'info': [10, 15, 27]})
test_dich3.update({'about': {'game': 'football', 'period': 5}})
test_dich3.update({'about': 'dont know'})
print(test_dich3)

{'info': [10, 15, 27], 'about': 'dont know'}


In [ ]:
s2 = {5, 10, 3, 2, 11}
s3 = set('wow thats cool')
s1 = set('abcde')
s1.add('hello')
s4 = {'a', 'b', 'c', 'd', 'e', 'f'}
s4.discard('f')
s4

{'a', 'b', 'c', 'd', 'e'}

In [ ]:
a = 1.3+1.3
a

2.6

In [ ]:
name_to_age = {'Anne': 22, 'Anton': 27, 'Phillip': 30}
print(name_to_age.keys())

dict_keys(['Anne', 'Anton', 'Phillip'])


In [ ]:
place_and_money = {1: 100, 2: 50, 3: 10}
print(place_and_money.values())

dict_values([100, 50, 10])


In [ ]:
#union
alpha_set = {'a', 'b', 'c', 'd', 'e'}
name = set('bad boy')

a = alpha_set | name
a

{' ', 'a', 'b', 'c', 'd', 'e', 'o', 'y'}

In [ ]:
#union
alpha_set = {'a', 'b', 'c', 'd', 'e'}
name = set('bad boy')
print(name)
a = alpha_set.union(name)
b = alpha_set | name
print(f"'union': {a} \n'  |  ': {b} ", a is b )

{'y', 'b', 'd', 'o', ' ', 'a'}
'union': {'c', 'y', 'b', 'd', 'o', ' ', 'a', 'e'} 
'  |  ': {'c', 'y', 'b', 'd', 'o', ' ', 'a', 'e'}  False


In [ ]:
num_set = set(range(0,11))
date_num = set([1,9,4,8])
a = num_set.union(date_num)
print(a)
print(num_set | date_num)

{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10}


In [ ]:
#intersection
alpha_set = {'a', 'b', 'c', 'd', 'e'}
b =set('abcde')
name = set('bad boy')
b.intersection(name), name.intersection(b)


({'a', 'b', 'd'}, {'a', 'b', 'd'})

In [ ]:
num_set = set(range(1,11))
date_num2 = set([1,9,4,8])
date_num2.intersection(num_set)


{1, 4, 8, 9}

In [ ]:
#difference
alpha_set = {'a', 'b', 'c', 'd', 'e'}
name = set('bad boy')
alpha_set.difference(name)

{'c', 'e'}

In [ ]:
num_set = set(range(0,11))
date_num2 = set([1,9,4,8])
num_set.difference(date_num2)

{0, 2, 3, 5, 6, 7, 10}

In [ ]:
#issubset


In [ ]:
num1 = 3
num2 = 3.5
sum1 = num1 + num2
a = float(num1)
sum2 = num2 + a
sum1, sum2

(6.5, 6.5)

In [ ]:
str_one = 1
str_one = str(str_one)
str_one

'1'

In [ ]:
str_age_begin = 'my age is '
int_age = 18
result_string = str_age_begin + str(int_age)
s = '11.1'
print(float(s))

11.1


In [ ]:
list6 = [3,1,-10,5,11,20,1,-10]
list7 = list6.copy()
list6.sort()
list_result = list6.copy()
print(list7, list_result)
list_result.reverse()
print(list7, list_result)

[3, 1, -10, 5, 11, 20, 1, -10] [-10, -10, 1, 1, 3, 5, 11, 20]
[3, 1, -10, 5, 11, 20, 1, -10] [20, 11, 5, 3, 1, 1, -10, -10]


In [ ]:
list1 = [10,15]
tpl1 = tuple(list1)
dict1 = {}
dict1[tpl1] = 'hello'
print(dict1)

{(10, 15): 'hello'}


In [ ]:
dict1 = {}
dict2 = dict()
dict1['name'] = 'unknown'
dict2['name'] = 'Tom'
print(dict1, dict2, dict1.keys(), dict1.values())

{'name': 'unknown'} {'name': 'Tom'} dict_keys(['name']) dict_values(['unknown'])


In [ ]:
s1 = set('hello')
s2 = set(['w', 'o', 'w'])
s3, s4 = s1.copy(), s1.copy()
s3.update(s2)
print(s4)
s4.discard(s2)
print(s4)
s4 -= s2
print(s4)
s1.intersection_update(s2)
print(s3, s1, s4)


{'h', 'l', 'o', 'e'}
{'h', 'l', 'o', 'e'}
{'h', 'l', 'e'}
{'h', 'w', 'l', 'o', 'e'} {'o'} {'h', 'l', 'e'}


In [ ]:
#  Условные операторы
# 2.4
target_string = "шалаш  "
target_string = (target_string.lower()).replace(' ', '')

result = target_string == target_string[::-1]
result

True

In [ ]:
# 2.8
#result = (a**2 > b) and (a*a > c)

In [ ]:
# 2.9
diagnosis_1 = 'yes'
diagnosis_2 = 'yes'
diagnosis_3 = 'no'

#result = diagnosis_1 +' '+ diagnosis_2 +' '+ diagnosis_3
result = diagnosis_1 == 'yes' or diagnosis_2 == 'yes' or diagnosis_3 == 'yes'
result

True

In [ ]:
# 3.1
N = 1700003000000000
contains = "3" in str(N) and '7' in str(N)
print(contains)


True


In [ ]:
#3.2
x, y = 'F', 9
o_x = x in set('ABCDEFGH')
o_y = str(y) in set('12345678')
move_is_correct = o_x and o_y
print(move_is_correct)

False


In [ ]:
# 3.3
N = 2700003000000000
first_is_even = int(str(N)[0]) % 2 == 0
print(first_is_even)

True


In [ ]:
# 3.4
year = 450
year_is_leap = \
(year % 400 == 0) or \
((year % 4 == 0) and \
(year % 100 != 0))
print(year_is_leap)


False


In [ ]:
# 3.5
n, m, k = 5, 4, 8 # ширина, длина, количество долек после разделения. отломленая часть
s = n * m
result = (((k % n == 0) and (n * (k // n) == k)) or \
((k % m == 0) and (m * (k // m) == k))) and s > k

In [ ]:
# 3.7
arrival_of_goods = {
    '148902': {
        'Футболка с принтом': 180,
        'Свитшот черный': 245,
        'Джинсы серые': 252
    },
    '893516': {
        'Футболка с принтом': 43,
        'Свитшот черный': 64,
        'Джинсы черные': 102
    },
    '893481': {
        'Кружка керамическая': 35,
        'Свитшот черный': 10,
        'Джинсы сервые': 14
    }
}
invoice_number = '893481'
# invoice_exists = invoice_number in set(arrival_of_goods.keys()) # работает но не по теме
invoice_exists = arrival_of_goods.get(invoice_number) is not None # если None то нет в словаре
print(invoice_exists)

True


In [ ]:
# *ТЕРНАРНЫЙ УСЛОВНЫЙ ОПЕРАТОР, тест
cust_age = 404
##if cust_age >= 60:
    # условие выполняется
    #print("Eligible for discount")
#else:
    # условие не выполняется
    #print("Not eligible for discount")

# № 2
# Реализуем тот же код через тернарный оператор
#cust_discount = "Eligible for discount" if cust_age >= 60 else "Not eligible for discount"
#print(cust_discount)

# мой вариант
print("Eligible for discount" if cust_age >= 60 else "Not eligible for discount")
# работает

Eligible for discount


In [ ]:
#5.3
height = 183
weight = 78
zodiac_sign = 'козерог'
zodiac_set = {'весы', 'дева', 'овен', 'козерог'}
if height > 180 and weight < 80 and zodiac_sign in zodiac_set:
    print('Ваша половинка нашлась!')
else:
    print('Попробуем поискать еще!')

Ваша половинка нашлась!


In [ ]:
# 5.5

player_1 = 'камень'
player_2 = 'ножницы'

if player_1 == player_2:
    print('Ничья!')
elif (player_1 == 'камень' and player_2 == 'ножницы') or \
    (player_1 == 'ножницы' and player_2 == 'бумага') or \
    (player_1 == 'бумага' and player_2 == 'камень'):
    print('Первый игрок — победитель!')
else:
    print('Второй игрок — победитель!')





Первый игрок — победитель!


In [ ]:
# 5.6

purchases = ["Adidas", 'Nike']
prices = {'Adidas': 4298, 'Nike': 6550, 'Puma': 4490, 'Asics': 3879}
# Ошибки (не обязательно в этой задаче, но чтобы отработать словарь)
#mistakes = {
#    'quantity': 'Ваша корзина пуста',
#}

if purchases:
#    mistakes.pop('quantity')
    if len(purchases) == 1:
        discount = 0
        print(f'Стоимость заказа составила: {prices[purchases[0]]}')
    else:
        discount = 10 if purchases[0] == purchases[1] else 5
        full_price = prices[purchases[0]]+prices[purchases[1]]
        print(f'Стоимость заказа составила: {full_price}. С учетом скидки в {discount}% — {full_price * (100-discount)/100}')
else:
    print('Ваша корзина пуста')





Стоимость заказа составила: 10848. С учетом скидки в 5% — 10305.6


In [ ]:
# 5.7

date = '16.04.2019 10:59'
str_date_list = date.split(' ')[0].split('.') # ['число', 'месяц', 'год']
date_dict = {'day': int(str_date_list[0]),
             'month': int(str_date_list[1]),
             'year': int(str_date_list[2])
             } # перевели все в числа

if date_dict['year'] == 2019:
    if date_dict['month'] == 5:
        category = 1
    elif 1 <= date_dict['month'] < 5:
        category = 2
elif date_dict['year'] < 2019:
    category = 3



In [ ]:
# 5.8
city_info = 'Волгоград , готов к переезду (Сочи, Москва) , готов к командировкам'

million_cities = [
    'Новосибирск', 'Екатеринбург', 'Нижний Новгород', 'Казань', \
    'Челябинск', 'Омск', 'Самара', 'Ростов-на-Дону', 'Уфа', \
    'Красноярск', 'Пермь', 'Воронеж', 'Волгоград']

city_category = ["Москва", "Санкт-Петербург", 'Город миллионник', 'Другое']
city_ = city_info.split(' , ')[0]

if (city_ == city_category[0]):
    city = city_category[0]
elif city_ == city_category[1]:
    city = city_category[1]
elif city_ in million_cities:
    city = city_category[2]
else:
    city = city_category[-1]


print(city)

Город миллионник


In [ ]:
# 6.5 try-except-else-finally

medicines = {'Ибупрофен': 99, 'Эспумизан': 279, 'Пенталгин': 119}
name = 'Ибупрофен'

try:
    print(medicines[name])
except:
    print('Такого ключа в словаре нет')


99


In [ ]:
# ЦИКЛЫ!!!!!!!!!!!!!

In [ ]:
# 2.5

#for i in list(range(1, n+1, 1)):
#    print('*'*i)
n = 5
k = [print('*'*i) for i in list(range(1, n+1, 1))] # УДАЛОСЬ ГЕНЕРАТОРОМ!!!
# ПОЧЕМУ ГЕНЕРАТОР ВЫДАЕТ ИМЕННО N - NONE если объявить без переменной???

TypeError: 'list' object is not callable

In [ ]:
# 2.9

num_list = [1, 10, 3, -5]
num_list.sort()
s=0
for num in num_list:
    print(f'element {s}: {num}')
    s += 1

In [ ]:
# 2.10

num_list = list(range(0, 100, 3))
count_even = 0
for num in num_list:
    if num % 2 == 0:
        count_even += 1


In [ ]:
# 2.11

mixture_list = [True, 1, -10, 'hello', False, 'string_1', 123, 2.5, [1, 2], 'another']
count_str = 0
for elem in mixture_list:
    if isinstance(elem, str):
        count_str += 1

In [ ]:
from io import TextIOBase
# 2.12

word_list = ["My", "name", "is", "Sergei", "EOS", "I'm", "from", "Moscow", "EOS"]
text = ''
for word in word_list:
    text += f'{word} '
#text = text.replace("EOS", '.').replace(' . ', '. ')
text = text.replace("EOS", '.').strip()
print(text)

In [ ]:
# 3.2 ЦИКЛЫ WHILE

volume = 15
H = 0 # объем воды в сосуде
h = 3.3 # литра
while H <= volume:
    H += h
cost = H - volume
print(cost)

In [ ]:
#3.3
a = 1
value = 10
while a**2 <= value:
    a += 1
    print(f'проверяю число = {a}')
print(f'\n{a}')

In [ ]:
# 3.9

x = 1
while (x ** 2 % n) != 0:
    x += 1
print(x)



In [ ]:
# 3.10

x = 0
p = 1
value = 1000
while True:
    p *= (x + 1)
    if p > value:
        break
    else:
        x += 1
print(p)

In [ ]:
# 3.11

money = 1000
target_money = 3000
year_count = 0
pr = 1.08
while money < target_money:
    money *= pr
    year_count += 1
print(year_count)

In [ ]:
# 3.12

health = 500
damage = 80
seconds_num = 0
while health > 0:
    health -= damage
    seconds_num += 1
print(seconds_num)

In [ ]:
# 3.13 Числа Фибоначчи для чисел n >= 2

n = 2
# fibonacci_list = [1, 1, 2, 3, 5]
fibonacci_list = [1, 1]
while n > 2:
    n -= 1
    fibonacci_list.append(fibonacci_list[-1]+fibonacci_list[-2])
print(fibonacci_list)

In [ ]:
# 4.4 вложенные списки. вложеные циклы for

random_matrix = [
    [9, 2, 1],
    [2, 5, 3],
    [4, 8, 5]
]
max_value_rows = []
for list in random_matrix:
    #print(list)
    max = list[0]
    for elem in list:
        if elem > max:
            max = elem
    max_value_rows.append(max)
print(max_value_rows)


In [ ]:
# 4.8 вложеные списки

test_matrix = [
    [1, 2, 3,  4],
    [7, -1, 2, 5],
    [123, 2, -1, 3],
    [123, 5, 1]
]
h = len(test_matrix)
for list in test_matrix:
    if len(list) != h:
        is_square = False
        break
    else:
        is_square = True
print(is_square)

In [ ]:
# 4.9 вложеные списки, температуру в +

temp = [[25, 27, 28, 26, 27, -26, -25, -2, 26],\
        [21, 22, 28, 27, 28, 26, 25, 19, 26], \
        [-19, 21, 25, -27, 28, 25, 21, 20, 26]
        ]
for list in temp:
    len_list = range(len(list))
    for i in len:
        if list[i] < 0:
            list[i] = - list[i]
print(temp)

# После обработки
## temp = [[25, 27, 28, 26, 27, 26, 25, 2, 26], [21, 22, 28, 27, 28, 26, 25, 19, 26], [19, 21, 25, 27, 28, 25, 21, 20, 26]]


In [ ]:
# 4.9 вложеные списки, температуру в +. Второй вариант

temp = [[25, 27, 28, 26, 27, -26, -25, -2, 26],\
        [21, 22, 28, 27, 28, 26, 25, 19, 26], \
        [-19, 21, 25, -27, 28, 25, 21, 20, 26]
]
for list in temp:
    list[:] = [-i if i<0 else i for i in list] # переписывается список
print(temp)

In [ ]:
# 4.10* Вложеные списки. Удовлетворенность клиентов 0-1

customer_satisfaction = [
    [0.87, 0.56, 0.77],
    [0.22, 0.46, 0.56, 0.89, 0.95],
    [0.45, 0.44, 0.68],
    [0.73, 0.88, 0.95, 0.49]
]
month_satisfaction = [] # среднее значение листа
max_satisfaction = float()

for list in customer_satisfaction:
    summ_satisfaction = float()
    count_elem = 0
    for elem in list:
        count_elem += 1
        summ_satisfaction += elem
        satisfaction = round(summ_satisfaction/count_elem, 2)
    month_satisfaction.append(satisfaction)
    if satisfaction > max_satisfaction:
        max_satisfaction = satisfaction

print(max_satisfaction)
print(month_satisfaction)


In [ ]:
# 5.3 самопроверка. enumerate()

user_dynamics = [-5, 2, 4, 8, 12, -7, 5]
number_negative = None #объявляем переменную, в которой будем хранить номер последнего дня оттока, изначально она пустая (None)
#создаём цикл по элементам последовательности от 0 до N (не включая N)
for index, value in enumerate(user_dynamics):
    if value < 0:
        number_negative = index + 1
        print("Churn value: ", value) # выводим количество ушедших в этот день пользователей
        print("Number day: ", number_negative) # выводим номер дня
#print('_-----------------БЫЛО - то что ниже--------------------_')
#user_dynamics = [-5, 2, 4, 8, 12, -7, 5]
#number_negative = None
#N = len(user_dynamics)
#for i in range(N): # i — индекс текущего элемента
    #проверяем условие оттока — текущий элемент отрицательный
#    if user_dynamics[i] < 0: #если условие истинно,
#        number_negative = i+1  #перезаписываем значение номера дня
#        print("Churn value: ", user_dynamics[i]) # выводим количество ушедших в этот день пользователей
#        print("Number day: ", number_negative) # выводим номер дня

In [ ]:
# 5.5

str_list = ['Hello', 'my', 'name', 'is', 'Ezeikel', 'I', 'like', 'knitting']
# cut_str_list = [[0, 'Hel'], [1, 'my'], [2, 'nam'], [3, 'is'], [4, 'Eze'], [5, 'I'], [6, 'lik'], [7, 'kni']]
cut_str_list = []
for index, value in enumerate(str_list):
    cut = [index, value[:3:]] # срез
    cut_str_list.append(cut)
print(cut_str_list)

In [ ]:
# 5.7
# Гипотеза Сиракуз заключается в том, что любое натуральное число n можно свести к 1, если повторять над ним следующие действия:
n = 19
origin = n
# Создаём бесконечный цикл

while True:
    # Ваш код здесь
    if origin % 2 == 0:
        origin = origin // 2



    # Проверяем, что результат равен 1
    if origin == 1:
        # Если условие выполняется, выводим утвердительное сообщение.
        print(f'Syracuse hypothesis holds for number {origin}')


        # Ваш код здесь
        break
    elif origin % 2 != 0:
        origin = (origin * 3 +1) // 2


In [ ]:
# 5.10 continue. сколько значений в словаре mixture_dict являются числами.

mixture_dict = {'a': 15, 'b': 10.5, 'c': '15', 'd': 50, 'e': 15, 'f': '15'}
count_numbers = 0
for kl in mixture_dict:
    if str(mixture_dict[kl]) is mixture_dict[kl]:
        continue
    else:
        count_numbers += 1
print(count_numbers)



In [ ]:
# 5.11 Разбить текст
text = """
*-+She sells sea shells on the sea shore;
The shells that she sells are sea shells I am sure.
So if she sells sea shells on the sea shore,
I am sure that the shells are sea shore shells.
"""

point = '.'
comma = ','
semicolon = ';'
question = '?'
exclamation = '!'
dash = '—'

count_punctuation = {point: 0, comma: 0, semicolon: 0, question: 0, exclamation: 0, dash: 0}
text = text.replace(' ', '').replace('\n', '')
for el in text:
    try:
        if not el.isalnum():
            count_punctuation[el] += 1
    except KeyError:
        continue
print(count_punctuation)

In [ ]:
# 5.12 Разбить текст

sentence = 'A roboT MAY Not injure a humAn BEING or, tHROugh INACtion, allow a human BEING to come to harm.'
sentence = sentence.replace('.', '').replace(',', '')
sentence = sentence.lower().split(' ')
word_dict = dict()
for el in sentence:
    if el not in word_dict:
        word_dict[el] = 1
    else:
        word_dict[el] += 1
print(word_dict)

In [ ]:
# 5.13 счет количества вхождений  символа в каждую из строк списка

str_list = ["text", "morning", "notepad", "television", "ornament"]
symbol_to_check = 't'
str_tuple = tuple(str_list)

word_dict = dict()  # удалять нельзя
for el in str_tuple:  # создаем словарь
    word_dict[el] = 0
    for i in el:
        if i == symbol_to_check:
            word_dict[el] += 1
print(word_dict)

In [ ]:
# 3.1 Функции/ работа с исключениясм
def get_time(distance, speed):
    if distance < 0 or speed < 0:
        raise ValueError("Distance or speed cannot be below 0!")
    if speed == 0:
        raise ValueError('Speed cannot be equal to 0!')
    result = distance / speed
    return result
get_time(100, 0)


In [ ]:
# 5.2 на самопроверку. РАСПАКОВКА *
langs = ['Python', 'SQL', 'Machine Learning', 'Statistics']
# Будет напечатано:                                при langs
# ['Python', 'SQL', 'Machine Learning', 'Statistics']
# Python SQL Machine Learning Statistics           при *langs
#for i in langs:
#    print(i, end= ', ')
print(*langs, sep=', ')

In [ ]:
def mean_mark(name, *marks):
    result = sum(marks) / len(marks)
    # Не возвращаем результат, а печатаем его
    print(name+':', result)
marks = [4,5,5,5,5,3,4,4,5,4,5]
mean_mark(marks, name = 'Kuznetsov')


In [ ]:
# ЛЯМБДА ФУНЦИЯ, пример с сортировкой
new_list = ['bbb', 'ababa','aaa', 'aaaaa',  'cc']
new_list.sort(key=lambda word: (len(word), word))
print(new_list)
# Будет напечатано:
# ['cc', 'aaa', 'bbb', 'aaaaa', 'ababa']

In [ ]:
# ЛЯМБДА ФУНЦИЯ, пример с сортировкой
def sort_sides(l_in):
   l_in.sort(key=lambda l_in: (l_in[0]**2 + l_in[1]**2 ) ** 1/2)
   return l_in

print(sort_sides(l_in = [(3, 4)]))


In [ ]:
# 7,9 повторение, можно ли лямбдой?

def get_less(list_in, num):
    for n in list_in:
        if num > n:
            return n


print(get_less([1, 5, 8,  10], 0))

In [ ]:
# 7.10

def split_date(date):
    d = int(date[:2])
    m = int(date[2:4])
    y = int(date[4::])
    return (d,m,y)
print(split_date("31012019"))
# (31, 1, 2019)

In [ ]:
# 7.11 ПРОВЕРКА ЧИСЛА НА ПРОСТОТУ

def is_prime(num):
    # Negative numbers, 0 and 1 are not primes
    if num > 1:

        # Iterate from 2 to n // 2
        for i in range(2, (num//2)+1):

            # If num is divisible by any number between
            # 2 and n / 2, it is not prime
            if (num % i) == 0:
                return False
                break
        else:
            return True
    else:
        return False
print(is_prime(13))

In [ ]:
# 7.12
def between_min_max(*nums):
    min, max = nums[0], nums[-1]
    for n in nums:
        if n > max:
            max = n
        elif n < min:
            min = n
    return (min+max)/2

print(between_min_max(10))
print(between_min_max(1,2,3,4,5))
# 10.0
# 3.0

In [ ]:
#  7.13
def  best_student(**dict_):
    min = None
    name = str()
    for i in dict_:
        if min == None:
            min = dict_[i]
            name = i
            continue
        elif dict_[i] < min:
            min = dict_[i]
            name = i
    return name



print(best_student(Tom = 12, Jerry = 1, Jane = 2))

In [ ]:
# 7.14 Лямбда функция
is_palindrom = lambda x : 'yes' if x == x[::-1] else 'no'

print(is_palindrom('12321'))
print(is_palindrom('1234'))

In [ ]:
# 7.15 Лямбда функция
area =lambda a,b : a * b
print(area(12,5))

In [ ]:
# 7.16... ИЗ 7.12 сделать ЛЯМБДУ

"""def between_min_max(*nums):
    min, max = nums[0], nums[-1]
    for n in nums:
        if n > max:
            max = n
        elif n < min:
            min = n
    return (min+max)/2

print(between_min_max(10))
print(between_min_max(1,2,3,4,5))
# 10.0
# 3.0"""

between_min_max = lambda *x : (min(x) + max(x)) / 2

print(between_min_max(1,2,3,4,5))
# 3.0


In [ ]:
# 7.17 функция сортировщик (2)
def sort_ignore_case(ls):
    ls.sort(key = lambda word: word.lower()) # ворд подается из параметра key
    return ls
#print(sort_ignore_case(['Acc', 'abc']))
# ['abc', 'Acc']

def equalize_lengths(ls):
    ls.sort(key = lambda word: -len(word)) # СОРТИРУЕТ ПО ДЛИНЕ СТРОКИ
    max_in_ls = max(map(lambda ls: len(ls), ls))
    func = lambda word: word + '_' * (max_in_ls-len(word)) if len(word) < max_in_ls else word
    lst = list(map(func, ls))
    return lst

print(equalize_lengths(['крот', 'белка', 'выхухоль']))
# ['выхухоль', 'белка', 'крот'] # ЗАДАЧА ДОЗАПОЛНИТЬ РАЗНИЦУ ДЛИН СТРОК

['выхухоль', 'белка___', 'крот____']


In [ ]:
# 7.18 функция высчитыватель
def exchange(usd = (None, 'usd'), rub = (None, 'rub'), rate = (None, 'rate')):
    count = 0
    given = []
    target = None
    for i in [usd, rub, rate]:
        if type(i) is tuple:
            target = i[1]
        else:
            count += 1
            given.append(i)
    if count == 3:
        raise ValueError('Too many arguments')
    elif count < 2:
        raise ValueError('Not enough arguments')
    if target == 'rate':
        result = given[1] / given[0]
    elif target == 'usd':
        result = given[0] / given[1]
    else:
        result = given[0] * given[1]
    return result

'''print(exchange(usd=100, rub=8500))
# 85.0
print(exchange(usd=100, rate=85))
# 8500
print(exchange(rub=1000, rate=85))
# 11.764705882352942

print(exchange(usd = 12, rub = 24, rate = 10))
print(exchange(usd = 12))'''

In [ ]:
a

In [ ]:
# 2.3 работа с функциями

def get_count_unique_symbols(text):
    text = text.lower().replace(' ', '')
    text_set = set(text)
    count_set = len(text_set)
    return count_set
get_count_unique_symbols('Это простая строка')
## 9
get_count_unique_symbols('This is a simple string')
## 12

In [ ]:
def get_min_string(s1, s2):
    def get_count_unique_symbols(text):
        text = text.lower().replace(' ', '')
        text_set = set(text)
        count_set = len(text_set)
        return count_set
    len_s1 = get_count_unique_symbols(s1)
    len_s2 = get_count_unique_symbols(s2)
    if len_s1 < len_s2:
        result = s1
    elif len_s1 == len_s2:
        result = (s1, s2)
    else:
        result = s2
    return result
print(get_min_string(s1='школа', s2='school'))
    ## ('школа', 'school')
print(get_min_string(s1='Отличная фраза', s2='Great phrase'))
    ## Great phrase

In [ ]:
# 2.9 Области видимости функций

pi = 3.1416
def calculate_area_circle(r):
    s = round(pi * r * r, 3)
    return s
def calculate_area_ellipse(a, b):
    s = round(pi * a * b, 3)
    return s
print(calculate_area_circle(r=5))
## 78.54
print(calculate_area_ellipse(a=3, b=2.5))
## 23.562

In [ ]:
# 3.2 Область видимости

def cash(less_money):

    global money
    money -= less_money
    return money

money = 200000
print(cash(1000))
## 199000

money = 30240
print(cash(240))
## 30000

In [ ]:
# 3.3 Область видимости. избавляемся от global

# Словарь с курсами валют (по отношению к рублю)
currencies = {'USD': 74, 'EUR': 88, 'GBP': 98 , 'CHF': 82}
# Общее количество денег на счету, которое нужно конвертировать
money = 100000
# Функция для конвертации валюты, аргумент - наименование валюты
def convert(currencies, money, val):
    # Объявляем, что money - глобальная переменная

    # Производим конвертацию - делим количество денег на счету на соответствующий курс
    money /= currencies[val]
    return money

# Вызываем функцию для конвертации валюты
print(convert(currencies, money, 'USD'))

In [ ]:
# 3.5 Область видимости

# Функция для вычисления количества символов (symbol) в строке s
def count_occurrences(s, symbol):
    # Внутренняя функция для предобработки строки s
    def preprocessing_s():
        nonlocal s
        # Удаляем пробелы из строки
        s = s.replace(' ', '')
        # Приводим строку к нижнему регистру
        s = s.lower()

    # Вызываем функцию для предобработки аргумента s
    preprocessing_s()
    # Считаем количество символов symbol в строке s и возвращаем результат

    return s.count(symbol)

count_occurrences('This is simple string', symbol='t')

In [ ]:
# 3.7 Область видимости
# Создаём новый пустой словарь
advertising_campaigns = {'ютуб': [212, 248], 'вк': [514, 342], 'радио': [339, 125]}

advertising_campaigns_max = {}
# Создаём цикл по ключам исходного словаря
for key in advertising_campaigns:
    # Вычисляем максимум в списке, лежащем по ключу key
    max9 = max(advertising_campaigns[key])
    # Добавляем максимум в новый словарь
    advertising_campaigns_max[key] = max9
print(advertising_campaigns_max)

In [ ]:

def check_date(day, month, year):
    def is_leap(year):
        if year % 400 == 0:
            return True
        elif year % 100 == 0:
            return False
        elif year % 4 == 0:
            return True
        else:
            return False

    # Проверяем день, месяц и год на целочисленность
    if (type(day) is not int) or (type(month) is not int) or (type(year) is not int):
        return False
    # Проверяем год на заданный диапазон
    if (year <= 1900) or (year >= 2022):
        return True
    # Проверяем месяц на заданный диапазон
    if (month < 1) or (month > 12):
        return False
    # Проверяем день на заданный диапазон
    if (day < 1) or (day > 31):
        return False
    # Проверяем апрель, июнь, сентябрь и ноябрь на количество дней
    if (month in [4,6,9,11]) and (day > 30):
        return False
    # Проверяем количество дней в феврале
    if not is_leap(year) and month == 2 and day > 28:
        return False
    elif is_leap(year) and month == 2 and day > 29:
        return False
    return True

print(check_date(29, 2, 2000))

In [ ]:
# 4.9 Практика

def circle(p1, p2):
    def radius(p1, p2):
        x1, y1 = p1
        x2, y2 = p2
        return ((x2-x1)**2+(y2-y1)**2)**0.5
    r = radius(p1,p2)
    def calculate_circumference():
        p = round(2* pi  * r, 3)
        return p
    l = calculate_circumference()
    def calculate_area_circle():
        s = round(pi  * r * r, 3)
        return s
    s = calculate_area_circle()
    r_ = round(r, 3)
    return {'radius': r_, 'circumference': l, 'area': s}

pi = 3.1416
print(circle(p1=(3, 2.5), p2=(4, 4.5)))
## {'radius': 2.236, 'circumference': 14.05, 'area': 15.708}

pi = 3.1416
print(circle(p1=(0, 0), p2=(1, 1)))
## {'radius': 1.414, 'circumference': 8.886, 'area': 6.283}


In [ ]:
#5.0# ПРАКТИКА

def ellipse(p1, p2, p3):
    def semi_axes():
        x1, y1 = p1
        x2, y2 = p2
        x3, y3 = p3
        a = ((x2-x1)**2+(y2-y1)**2)**0.5
        b = ((x3-x1)**2+(y3-y1)**2)**0.5
        return a, b
    a, b = semi_axes()
    def calculate_area_ellipse():
        s = pi*a*b
        return round(s,3)
    s = calculate_area_ellipse()
    def calculate_length_ellipse():
        l = 2 *pi *((a**2+b**2)/2)**0.5
        return round(l, 3)
    l = calculate_length_ellipse()
    a_ = round(a,3)
    b_ = round(b, 3)
    return {'a': a_, 'b': b_, 'length': l, 'area': s}

pi = 3.1416
print(ellipse(p1=(3, 2.5), p2=(4.5, 2.5), p3=(3, 3.5)))

## {'a': 1.5, 'b': 1.0, 'length': 8.01, 'area': 4.712}

pi = 3.1416
print(ellipse(p1=(0, 0), p2=(0, 1), p3=(1, 0)))

## {'a': 1.0, 'b': 1.0, 'length': 6.283, 'area': 3.142}

pi = 3.14
print(ellipse(p1=(0, 0), p2=(0, 1), p3=(1, 0)))
## {'a': 1.0, 'b': 1.0, 'length': 6.28, 'area': 3.14}

In [ ]:
# 5.5 РЕКУРСИЯ

def inv_sum_list(lst):
    if not lst:
        return 0

    return 1/lst[0] + inv_sum_list(lst[1::])

print(inv_sum_list([10, 4, 8]))
## 0.475

In [ ]:
# Закрепление рекурсии. ФАКТОРИАЛ. СОЧЕТАНИЕ "C=n!/((n-k)!*k!)"

def combination(n, k):
    def factorial(x):
        if x == 0: return 1
        if x == 1: return 1
        return x * factorial(x-1)
    result = factorial(n)/ factorial(n-k) / factorial(k)
    return result

print(combination(n=10, k=5))
## 252.0
print(combination(n=12, k=3))
## 220.0
print(combination(n=1, k=1))
## 1.0
print(combination(n=0, k=0))
## 1.0

In [ ]:
# РЕКУРСИЯ. Число фибоначи

def fib(n):
    if n == 1: return 1
    if n == 2: return 1
    return fib(n-1)+fib(n-2)

In [ ]:
# РЕКУРСИЯ. ЗАДАЧА ДЛЯ СОБЕСЕДОВАНИЯ

def add_brackets(s):
    # Проверяем условие остановки: строка состоит из одного или двух символов
    if len(s) == 1 or len(s) == 2:
        # Возвращаем эти символы
        return s
    # В противном случае:
    # «Отщипываем» от строки первый и последний символы,
    # добавляем к ним скобки, а также результат вызова функции, в которую
    # передаем строку без первого и последнего символов
    return s[0] + '(' + add_brackets(s[1:-1]) + ')' + s[-1]
print(add_brackets('example'))
print(add_brackets('carr'))
print(add_brackets('hello'))

## Будет выведено:
## e(x(a(m)p)l)e
## c(ar)r
## h(e(l)l)o

In [ ]:
#РЕКУРСИЯ. СЛОЖНАЯ ЗАДАЧКА
def add_asterisk(s):
    if len(s) == 1:
        return s
    if len(s) == 2:
        return s[0] + '*' + s[-1]
    return s[0] + '*' + add_asterisk(s[1:-1]) + '*' + s[-1]

print(add_asterisk('hello'))
## h*e*l*l*o
print(add_asterisk('LItBeoFLcSGBOFQxMHoIuDDWcqcVgkcRoAeocXO'))
## L*I*t*B*e*o*F*L*c*S*G*B*O*F*Q*x*M*H*o*I*u*D*D*W*c*q*c*V*g*k*c*R*o*A*e*o*c*X*O
print(add_asterisk('gkafkafkKdaflkfa'))
## g*k*a*f*k*a*f*k*K*d*a*f*l*k*f*a

In [ ]:
# ПИЗДЕЦ А НЕ ЗАДАЧА. КРУТАЯ РЕКУРСИЯ!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

# Функция для выпрямления списка
def flatten(lst):
    # Создаём новый пустой список
    result = []
    # Создаём цикл по элементам списка
    for elem in lst:
        # Если элемент списка является списком,
        if type(elem) is list:
            # Применяем к нему функцию выпрямления и добавляем элементы к результату
            result += flatten(elem)
        else: # Если элемент не является списоком,
            # Добавляем элемент в новый список
            result.append(elem)
    return result
matrix = [
    [1, 1, 0],
    [4, 2, 1],
    [0, 2, 1]
]
print(flatten(matrix))
## Будет выведено:
## [1, 1, 0, 4, 2, 1, 0, 2, 1]

matrix = [
    [1, 1, [5, 10, 34, [24, 1, 0]]],
    [4, [9, 8, 10], 1],
    [0, 2, 1]
]
print(flatten(matrix))
## Будет выведено:
## [1, 1, 5, 10, 34, 24, 1, 0, 4, 9, 8, 10, 1, 0, 2, 1]



In [ ]:
# РЕКУРСИЯ/подобная задача на сумму матрицы

def sum_list(n):
    sum = 0
    for i in n:
        if type(i) is list:
            sum += sum_list(i)
        else:
            sum += i
    return sum

matrix = [
    [1, 1, 0],
    [4, 2, 1],
    [0, 2, 1]
]
print(sum_list(matrix))
## 12

matrix = [
    [1, 1, [1, 2, 3], 0],
    [4, 2, 1, [10, 52, 2]],
    [0, 2, 1]
]
print(sum_list(matrix))
## 82

In [ ]:

forum_messages = {
    1: {'parrent_link': None, 'child_link': [3, 4]},
    2: {'parrent_link': None, 'child_link': [5]},
    3: {'parrent_link': 1, 'child_link': [6]},
    4: {'parrent_link': 1, 'child_link': []},
    5: {'parrent_link': 2, 'child_link': []},
    6: {'parrent_link': 3, 'child_link': []}
}
# Функция для удаления сообщения на форуме и всех его потомков
def delete_message(messages, msg_id):
    # Удаляем из словаря сообщение с идентификатором msg_id
    # Метод pop() возвращает значение, лежащее по удаляемому ключу
    result = messages.pop(msg_id)
    # Получаем идентификатор родителя
    parrent_link = result['parrent_link']
    # Получаем список идентификаторов потомков
    child_link = result['child_link']
    # Если у сообщения был родитель и он ещё не был удален
    # Эта запись будет аналогична parrent_link is not None
    if parrent_link and parrent_link in messages:
        # Обращаемся к словарю messages по ключу родителя
        # Удаляем потомка из списка потомков
        messages[parrent_link]['child_link'].remove(msg_id)
    # Если у сообщения были потомки
    # Эта запись будет аналогична child_link == [<значение_1>, <значение_2>, ...]
    if child_link:
        # В цикле проходимся по всем потомкам
        for child_id in child_link:
            # И повторяем те же самые действия для каждого из них
            # (рекурсивно вызываем функцию delete_message)
            delete_message(messages, msg_id=child_id)
    return messages

#Для сообщения, у которого есть родитель, но нет потомков:
print(delete_message(forum_messages, msg_id=5))
## Будет выведено:
## {1: {'parrent_link': None, 'child_link': [3, 4]}, 2: {'parrent_link': None, 'child_link': []}, 3: {'parrent_link': 1, 'child_link': [6]}, 4: {'parrent_link': 1, 'child_link': []}, 6: {'parrent_link': 3, 'child_link': []}}

#Для сообщения, у которого есть родитель и есть потомки:
#print(delete_message(forum_messages, msg_id=3))
## Будет выведено:
## {1: {'parrent_link': None, 'child_link': [4]}, 2: {'parrent_link': None, 'child_link': [5]}, 4: {'parrent_link': 1, 'child_link': []}, 5: {'parrent_link': 2, 'child_link': []}}
"""запускать поочередно, тк словарь - измениемый тип, 2 запуска подряд удаляют одновременно """

In [ ]:
# 6.8 нужно было найти ошибку level+1 добавили 1

input_dict = {
    'key1': {
        'key2': ['value1', 'value2'],
        'key3': {
            'key4': ['value3']
        }
    },
    'key5': {
        'key6': {
            'key7': ['value3', 'value5', 'value6']
        }
    }
}
def print_dict(input_data, level=0):
    # Если input_data — словарь

    if type(input_data) is dict:
        # Создаём цикл по ключам словаря

        for key in input_data:
            # Выводим ключ в формате "<пробелы> <имя ключа> ->"
            print('  ' * level + '{} ->'.format(key))
            # Повторяем те же операции для каждого значения словаря

            print_dict(input_data[key], level+1)

    else: # В противном случае
        # Выводим значения в формате "<пробелы> <значения>"

        print('  ' * level + str(input_data))


print_dict(input_dict)

In [ ]:
# 7.3 ФУНКЦИИ map

prices = {'яблоко': 99, 'апельсин': 99, 'вишня': 147, 'персик': 145, 'грейпфрут': 139}
## new_prices = {'яблоко': 94.05, 'апельсин': 94.05, 'вишня': 139.65, 'персик': 137.75, 'грейпфрут': 132.05}
def pre_func(prices_items):

    name, q = prices_items
    return name, round(q*0.95, 2)

new_prices = dict(map(pre_func, prices.items()))
print(new_prices)



{'яблоко': 94.05, 'апельсин': 94.05, 'вишня': 139.65, 'персик': 137.75, 'грейпфрут': 132.05}


In [ ]:
# 7.6 совместное использование мap b lambda

data = [(0.00632, 6.575, 65.2, 296.0, 4.98),
(0.02731, 6.421, 78.9, 242.0, 9.14),
(0.02729, 7.185, 61.1, 242.0, 4.03),
(0.03237, 6.998, 45.8, 222.0, 2.94),
(0.06905, 7.147, 54.2, 222.0, 5.33),
(0.02985, 6.43, 58.7, 222.0, 5.21),
(0.08829, 6.012, 66.6, 311.0, 12.43)]

updated_data = list(map(lambda x: (*x, round(x[0]*x[3]*x[4], 2)), data)) # ВНИМАНИЕ НА ЛЯМБДУ, МЫ ПЕРЕПИСЫВАЕМ КОРТЕЖ
print(updated_data)

[(0.00632, 6.575, 65.2, 296.0, 4.98, 9.32), (0.02731, 6.421, 78.9, 242.0, 9.14, 60.41), (0.02729, 7.185, 61.1, 242.0, 4.03, 26.61), (0.03237, 6.998, 45.8, 222.0, 2.94, 21.13), (0.06905, 7.147, 54.2, 222.0, 5.33, 81.7), (0.02985, 6.43, 58.7, 222.0, 5.21, 34.53), (0.08829, 6.012, 66.6, 311.0, 12.43, 341.31)]


In [ ]:
# ПОЯСНЕНИЯ ПО ФУНКЦИИ FILTER
# filter(<имя_функции>, <итерируемый_объект>) ФУНКЦИЯ ДОЛЖНА ВОЗВРАЩАТЬ    TRUE or FALSE

words_list = ["We're", 'in', 'a', 'small', 'village', 'near', 'Chicago', 'My', "cousin's", 'getting', 'married.']

def is_even(x):
    return len(x) % 2 == 0
even_list = list(filter(is_even, words_list))
# Смотрим, что получилось
print(even_list)
## Будет выведено:
## ['in', 'near', 'My', "cousin's", 'married.']

# Применяем lambda-функцию к каждому элементу списка
even_list = filter(lambda x: len(x) % 2 == 0, words_list)
# Смотрим, что получилось
print(list(even_list))

## Будет выведено:
## ['in', 'near', 'My', "cousin's", 'married.']

str_tuple = ("Москва", "15.1 см", "зацвело", "было пол 5 утра", "рассвет")
filtered_tuple = tuple(filter(lambda x: x.isalpha(), str_tuple))
# Смотрим, что получилось
print(filtered_tuple)

## Будет выведено:
## ('Москва', 'зацвело', 'рассвет')
###################################
data = [
    ("FPW-2.0_D", "Бонус: Тренажер по HTML", 10, 100, 10),
    ("FPW-2.0", "Бонус: Тренажер по JavaScript", 9.2, 70, 180),
    ("FPW-2.0_D", "Бонус: Тренажер по React", 8.5, 66.67, 68),
    ("FPW-2.0", "Бонусный: IT в современном мире", 8.64, 83.74, 856),
    ("FPW-2.0", "Бонусный: Введение", 8.73, 56.24, 745),
    ("FPW-2.0", "Бонус: D1. Знакомство с Django (NEW)", 9.76, 95.24, 21),
    ("FPW-2.0_D", "Бонус: D2. Модели (NEW)", 9.44, 77.78, 18)
]
#Как всегда, для решения задачи нам необходимо выделить в отдельную функцию действия, которые мы должны произвести с одним элементом итерируемого объекта (у нас это список)
#Поэтому для начала создадим функцию, которая будет принимать один кортеж с данными о модуле
def filter_module(module):
    # Распаковываем кортеж на пять переменных
    code, name, avg_votes, nessa, count = module
    # Создаём условия
    cond_1 = code == "FPW-2.0"
    cond_2 = nessa >= 70
    cond_3 = count > 50
    # Условия должны выполняться одновременно
    return cond_1 and cond_2 and cond_3

print(filter_module(module=("FPW-2.0", "Бонус: Тренажер по JavaScript", 9.2, 70, 180)))
## Будет выведено:
## True ->  Функция работает

filtered_data = list(filter(filter_module, data))
print(filtered_data)
## Будет выведено:
## [('FPW-2.0', 'Бонус: Тренажер по JavaScript', 9.2, 70, 180), ('FPW-2.0', 'Бонусный: IT в современном мире', 8.64, 83.74, 856)]
#функцию можно переписать лямбдой:
lambda_filter_module = lambda x: (x[0] == "FPW-2.0") and (x[3] >= 70) and (x[4] > 50)

In [ ]:
# 8.1 практика по фильтрам

prices = [34562, 66572, 25683, 17683, 56389, 28973]
## filtered_prices = [25683, 17683, 28973]
func = lambda x: x <= 30000
filtered_prices = list(filter(func, prices))
print(filtered_prices)

prices_1 = [21490, 24901, 24901, 34901, 24142, 64521]
filtered_prices_1 = list(filter(func, prices_1))
## filtered_prices = [21490, 24901, 24901, 24142]
print(filtered_prices_1)

[25683, 17683, 28973]
[21490, 24901, 24901, 24142]


In [ ]:
## 8.2 практика по фильтрам. отфильтровать входной список

family_list = [
    'certificate of a large family',
    'social card',
    'maternity capital',
    'parking permit',
    'tax benefit',
    'reimbursement of expenses',
    "compensation for the purchase of children's goods"
 ]
#"сертификат многодетной семьи",
#"социальная карта",
#"материнский капитал",
#"разрешение на парковку",
#"налоговая льгота",
#"возмещение расходов",
#"компенсация на покупку детских товаров"
def filter_family(benefit):
    return list(filter(lambda benefit: benefit in family_list, benefit))

print(filter_family(['newborn registration', 'parking permit',
                     'maternity capital', 'tax benefit', 'medical policy']))
## ['parking permit', 'maternity capital', 'tax benefit']

print(filter_family(['reimbursement of expenses', 'parking permit', 'maternity capital', 'medical policy']))
## ['reimbursement of expenses', 'parking permit', 'maternity capital']


In [ ]:
# ПРАКТИКА ПО ФИЛЬТРАМ

# Создаём пустой список, в который будем добавлять результаты
count_a = []
# Создаём цикл по элементам списка words_list
for word in words_list:
    # Проверяем условие, что длина имени больше либо равна пяти
    if len(word) >= 5:
        # Создаём кортеж (слово, количество букв 'a')
        result_tuple = (word, word.lower().count('a'))
        # Добавляем в итоговый список кортеж
        count_a.append(result_tuple)
print(count_a)

## Будет выведено:
## [("We're", 0), ('small', 1), ('village', 1), ('Chicago', 1), ("cousin's", 0), ('getting', 0), ('married.', 1)]

# Отлично, всё работает так, как мы задумывали. Но можно ли решить эту задачу проще?

# Отбираем слова из пяти и более букв
filtered_words = filter(lambda x: len(x) >= 5, words_list)
# Все отобранные слова переводим в нижний регистр и считаем число букв 'a' в них
# Результат выдаём в виде кортежа (слово, количество букв "a")
count_a = map(lambda x: (x, x.lower().count('a')), filtered_words)
# Переводим объект map в list и печатаем его
print(list(count_a))

## Будет выведено:
## [("We're", 0), ('small', 1), ('village', 1), ('Chicago', 1), ("cousin's", 0), ('getting', 0), ('married.', 1)]

Когда мы будем более детально изучать итераторы, мы поймём, что при создании итератора на самом деле не происходит самих вычислений. Вычисления происходят только на этапе распаковки итератора в список с помощью функции list().

Это значит, что вы не тратите время на промежуточную распаковку содержимого итератора (объекта map или filter) в список. Распаковка осуществляется только один раз — в момент вызова строки list(count_a). Таким образом немного повышается производительность кода, что может дать существенный бонус при работе с огромными массивами данных, состоящими из миллионов записей.

Нам необходимо выбрать из данных только тех пациентов, у кого соотношение массы и роста находится в пределах нормы.

Результатом работы нашей программы должен стать список кортежей, удовлетворяющих условию.

Например, для приведённого выше списка data мы должны получить следующий список:

[('Amanda', 1.61, 51, 19.7), ('Patricia', 1.65, 61, 22.4), ('Andrey', 1.79, 61, 19.0), ('Felicia', 1.63, 56, 21.1)]
РЕШЕНИЕ

Вычислять индекс массы тела мы уже умеем (вспомните задачу 5 из предыдущего юнита):

# Создаём lambda-функцию, которая считает BMI, и применяем её к каждому элементу списка
map_func = lambda x: (*x, x[2]/(x[1]**2))
updated_data = map(map_func, data)
Мы получили объект map и записали его в переменную updated_data. Давайте теперь напишем функцию для фильтрации.

Как мы уже обсуждали, выходом map() является кортеж из четырёх элементов (индексы идут от 0 до 3), где элемент под индексом 3 — это интересующий нас BMI.

По условию задачи, мы должны отобрать только тех, у кого BMI находится в пределах нормы (от 18.5 до 25 включительно). Давайте создадим lambda-функцию, которая будет принимать на вход кортеж и возвращать True, если элемент под индексом 3 (BMI) будет находиться в заданных пределах. Применим эту функцию к результату работы map() с помощью функции filter() и обернём результат в список.

Итоговый код будет выглядеть следующим образом:

# Создаём lambda-функцию, которая считает BMI, и применяем её к каждому элементу списка
map_func = lambda x: (*x, x[2]/(x[1]**2))
updated_data = map(map_func, data)
# Создаём lambda-функцию, которая возвращает True, если 18.5 <= BMI <= 25
filter_func = lambda x: 18.5 <= x[3] <= 25
# Применяем эту функцию к результату работы map()
filtered_data = filter(filter_func, updated_data)
# Переводим объект filter в list и печатаем его
print(list(filtered_data))

## Будет выведено:
## [('Amanda', 1.61, 51, 19.7), ('Patricia', 1.65, 61, 22.4), ('Andrey', 1.79, 61, 19.0), ('Felicia', 1.63, 56, 21.1)]

In [ ]:
# 8.3 ФИЛЬТРЫ

reg = [('Ivanov', 'Sergej', 24, 9, 1995),
      ('Smith', 'John', 13, 2, 2003),
      ('Petrova', 'Maria', 13, 3, 2003)]
filter_func = lambda x: x[4] > 2000
union_func = lambda x: (x[0] + ' ' + x[1][0]+'.', x[2], x[3], x[4])
result = filter(filter_func, reg)
new_reg  = list(map(union_func, result))
print(new_reg)

#или \
reg_ = [('Ivanov', 'Sergej', 24, 9, 1995),
      ('Smith', 'John', 13, 2, 2003),
      ('Petrova', 'Maria', 13, 3, 2003)]
filter_func_ = lambda x: x[4] > 2000
union_func_ = lambda x: (x[0] + ' ' + x[1][0]+'.', x[2], x[3], x[4])
new_reg_  = list(map(union_func, filter(filter_func_, reg_)))
print(new_reg_)

[('Smith J.', 13, 2, 2003), ('Petrova M.', 13, 3, 2003)]
[('Smith J.', 13, 2, 2003), ('Petrova M.', 13, 3, 2003)]


In [ ]:
## Введите свое решение ниже
data = [(0.00632, 6.575, 65.2, 296.0, 4.98),
 (0.02731, 6.421, 78.9, 242.0, 9.14),
 (0.02729, 7.185, 61.1, 242.0, 4.03),
 (0.03237, 6.998, 45.8, 222.0, 2.94),
 (0.06905, 7.147, 54.2, 222.0, 5.33),
 (0.02985, 6.43, 58.7, 222.0, 5.21),
 (0.08829, 6.012, 66.6, 311.0, 12.43)]

append_func = lambda x: (*x, round(x[0]*x[3]*x[4], 2))
filter_func= lambda x: x[5] > 60
filtered_data = list(filter(filter_func, map(append_func, data)))
print(filtered_data)

[(0.02731, 6.421, 78.9, 242.0, 9.14, 60.41), (0.06905, 7.147, 54.2, 222.0, 5.33, 81.7), (0.08829, 6.012, 66.6, 311.0, 12.43, 341.31)]


In [ ]:
## Введите свое решение ниже
input_string = 'Hello! My name is Python. I will help you to analyze some data.'
## count_words = 13

input_string = 'There are many great articles about Artificial Intelligence and its benefits for business and society. However, many of these articles are too technical for the average reader.'
## count_words = 27

count_words = len(''.join(input_string)\
                  .replace('.','').replace(',','')\
                  .replace('!','').replace('?', '').split())
print(count_words)

27


In [ ]:
## Введите свое решение ниже
file_path = 'data/images/train/10394.jpg'
## file_name = '10394'
## file_extension = 'jpg'
file_name, file_extension = file_path.split('/')[-1].split('.')

print(file_name, file_extension)

10394 jpg


In [ ]:
# Практика, повторение. !!!!!!!!!!!!!!!!можно ли решить проще?

#generated_text = "глаза нее на поднял он и она попросила что-нибудь скажи"
## updated_text = "скажи что-нибудь попросила она и он поднял на нее глаза"

generated_text = "задачи своей решения способ или информацию ищет он поисковик в запрос вводит человек когда"
## updated_text = когда человек вводит запрос в поисковик он ищет информацию или способ решения своей задачи
def revers_func(text):
    new_text = []
    a = text.split()
    l = len(a)-1
    for i in range(0, l+1):
        new_text.append(a[l-i])
    return " ".join(new_text)
updated_text = revers_func(generated_text)
print(revers_func(generated_text))


когда человек вводит запрос в поисковик он ищет информацию или способ решения своей задачи <class 'str'>


In [ ]:
## Введите свое решение ниже
car_dict = {
    'car_ID': [123, 117, 111, 82, 101, 96, 156, 2, 58, 49],
    'fueltype': ['gas', 'diesel', 'diesel', 'gas', 'gas', 'gas', 'gas', 'gas', 'gas', 'gas'],
    'horsepower': [68, 95, 95, 88, 97, 69, 62, 111, 101, 176],
    'price': [7609.0, 17950.0, 13860.0, 8499.0, 9549.0, 7799.0, 8778.0, 16500.0, 13645.0, 35550.0]
}
mean_price = sum(car_dict['price'])/len(car_dict['price'])

count_diesel = car_dict['fueltype'].count('diesel')

min_horsepower = min(car_dict['horsepower'])

13973.9
2


In [ ]:
# Введите свое решение ниже
def division(a, b):
    try:
        if b == 0:
            raise ZeroDivisionError("Zero division error!")
        div = a/b
    except ZeroDivisionError:
        return "Zero division error!"
    return div
division(a = 36, b = 0)

'Zero division error!'

In [ ]:
data = {'0': 1, '4': 1, '6': 1, '8': 2, '9': 1}

def holes_count(number):
    n_s = str(number)
    sum_ = 0
    for i in n_s:
        if i in data.keys():
            sum_ += data[i]
    return sum_

print(holes_count(8))
# 2
print(holes_count(146))

2
2


In [ ]:
# Ищем КОЛЛИЧЕСТВО ЧЕТНЫХ ЧИСЕЛ В МАТРИЦЕ

def even_numbers_in_matrix(matrix):
    result = []
    for list_ in matrix:
        result =result + list_
    new_list = list(filter(lambda l: l%2==0, result))
    return len(new_list)




matrix_example = [
    [1, 5, 4],
    [4, 2, -2],
    [7, 65, 88]
]

print(even_numbers_in_matrix(matrix=matrix_example))
# 5

5


In [ ]:
# СУМИРУЕМ МАТРИЦЫ(напролом)

def matrix_sum(matrix1, matrix2):
    result = []
    c1 = []
    for list_ in matrix1:
        c1.append(len(list_))
        result =result + list_
    result_2 = []
    c2 = []
    for list_ in matrix2:
        c2.append(len(list_))
        result_2 =result_2 + list_
    if c1 != c2:
        print("Error! Matrices dimensions are different!")
        return None
    new_list = []
    result_ =[]
    for i, val in enumerate(result_2):
        result_.append(result[i]+result_2[i])
        if (i+1)%3 ==0 and i+1 != 0 :
            new_list.append(result_)
            result_ =[]
    return new_list



matrix1_example = [
          [1, 5, 2],
          [4, 2, -2],
          [4, 2, -2]
]

matrix2_example = [
          [10, 15, 43],
          [41, 2, -2],
          [7, 5, 7]
]
print(matrix_sum(matrix1_example, matrix2_example))

Error! Matrices dimensions are different!
None


In [ ]:
## Выводим случайные данные по алфавиту

def print_personal_data(**arg):
    dict_= dict(sorted(arg.items()))
    for key in dict_:
        print(f"{key}: " + f'{dict_[key]}')




print_personal_data(first_name='John', last_name='Doe', age=28, position='Python developer')
# age: 28
# first_name: John
# last_name: Doe
# position: Python developer

#print_personal_data(first_name='Jack', last_name='Smith', age=32, work_experience = '5 years', position='Project manager')
# age: 32
# first_name: Jack
# last_name: Smith
# position: Project manager
# work_experience: 5 years


age: 28
first_name: John
last_name: Doe
position: Python developer


In [ ]:
# ФУНКЦИЯ ОЧИСТКИ ОТ ЗНАКОВ ПРЕПИНАНИЯ

def get_words_list(text):
    punctuation_list = ['.', ',', ';', ':', '...', '!', '?', '-', '"', '(', ')']
    txt = "".join(filter(lambda s: s if s not in punctuation_list else "" ,text.lower() ))
    return txt.split()

text_example = "Arrakis, the planet known as Dune, is forever his place."

print(get_words_list(text=text_example))
# ['arrakis', 'the', 'planet', 'known', 'as', 'dune', 'is', 'forever', 'his', 'place']


['arrakis', 'the', 'planet', 'known', 'as', 'dune', 'is', 'forever', 'his', 'place']


In [ ]:
# ФУНКЦИЯ ДЛЯ ПОИСКА УНИКАЛЬНЫХ СЛОВ(возвращает отсортироваными)

def get_unique_words(words_list ):
    words_set = set(words_list)
    words_list_to = sorted(list(words_set))
    return words_list_to

words_list_example = ['and', 'take', 'the', 'most', 'special', 'care', 'that', 'you', 'locate', "muad'dib", 'in', 'his', 'place', 'the', 'planet', 'arrakis', 'do', 'not', 'be', 'deceived', 'by', 'the', 'fact', 'that', 'he', 'was', 'born', 'on', 'caladan', 'and', 'lived', 'his', 'first', 'fifteen', 'years', 'there', 'arrakis', 'the', 'planet', 'known', 'as', 'dune', 'is', 'forever', 'his', 'place']

print(get_unique_words(words_list=words_list_example))
## ['and', 'arrakis', 'as', 'be', 'born', 'by', 'caladan', 'care', 'deceived', 'do', 'dune', 'fact', 'fifteen', 'first', 'forever', 'he', 'his', 'in', 'is', 'known', 'lived', 'locate', 'most', "muad'dib", 'not', 'on', 'place', 'planet', 'special', 'take', 'that', 'the', 'there', 'was', 'years', 'you']


['and', 'arrakis', 'as', 'be', 'born', 'by', 'caladan', 'care', 'deceived', 'do', 'dune', 'fact', 'fifteen', 'first', 'forever', 'he', 'his', 'in', 'is', 'known', 'lived', 'locate', 'most', "muad'dib", 'not', 'on', 'place', 'planet', 'special', 'take', 'that', 'the', 'there', 'was', 'years', 'you']


ФУНКЦИЯ СОРТИРОВЩИК - ВЫВОДИТ САМОЕ УПОМИНАЕМОЕ СЛОВО


In [ ]:
def get_most_frequent_word(text):

    def get_words_list(text):  # ФУНКЦИЯ ОЧИСТКИ ОТ ЗНАКОВ ПРЕПИНАНИЯ
        punctuation_list = ['.', ',', ';', ':', '...', '!', '?', '-', '"', '(', ')']
        txt = "".join(filter(lambda s: s if s not in punctuation_list else "" ,text.lower() ))
        return txt.split()

    def get_unique_words(words_list ): # ФУНКЦИЯ ДЛЯ ПОИСКА УНИКАЛЬНЫХ СЛОВ(возвращает отсортироваными)
        words_set = set(words_list)
        words_list_to = sorted(list(words_set))
        return words_list_to
    unique_words = get_unique_words(get_words_list(text)) # список уникальных слов

    frequent_words_set = dict()
    for word in unique_words:
        frequent_words_set[word] = text.split().count(word)
    if not unique_words:
        return str()
    return sorted(frequent_words_set.items(), key=lambda item: item[1])[-1][0]

#text_example = "A beginning is the time for taking the most delicate care that the balances are correct. This every sister of the Bene Gesserit knows. To begin your study of the life of Muad'Dib, then take care that you first place him in his time: born in the 57th year of the Padishah Emperor, Shaddam IV. And take the most special care that you locate Muad'Dib in his place: the planet Arrakis. Do not be deceived by the fact that he was born on Caladan and lived his first fifteen years there. Arrakis, the planet known as Dune, is forever his place."
#print(get_most_frequent_word(text_example))
# the
text_example = "dddd ddddf gh ddddf"
print(get_most_frequent_word(text_example))
# и

ddddf


ФУНКЦИЯ добавляет в каждый кортеж новый элемент + сортирует результат по добавленному элементу

In [ ]:
data = [
    ('Amanda', 37, 78, 67),
    ('Patricia', 78, 93, 68),
    ('Marcos', 79, 67, 89),
    ('Dmitry', 67, 68, 100),
    ('Andrey', 100, 78, 76),
    ('Victoria', 93, 69, 96),
]
def get_total_score(data):
    append_func = lambda x: (*x, x[1]+x[2]+x[3])
    data_ = sorted(list(map(append_func, data)), key = lambda x: -x[-1]) # реверс заменили минусом
    return data_





print(get_total_score(data))
# [('Amanda', 37, 78, 67, 182), ('Marcos', 79, 67, 89, 235), ('Dmitry', 67, 68, 100, 235), ('Patricia', 78, 93, 68, 239), ('Andrey', 100, 78, 76, 254), ('Victoria', 93, 69, 96, 258)]


[('Victoria', 93, 69, 96, 258), ('Andrey', 100, 78, 76, 254), ('Patricia', 78, 93, 68, 239), ('Marcos', 79, 67, 89, 235), ('Dmitry', 67, 68, 100, 235), ('Amanda', 37, 78, 67, 182)]


In [ ]:
data = {
    "type": "video",
    #"videoID": "vid001",
    "links": [
        {"type":"video", "videoID":"vid002", "links":[]},
        {   "type":"video",
            "videoID":"vid003",
            "links": [
            {"type": "video", "videoID":"vid004"},
            {"type": "video", "videoID":"vid005"},
            ]
        },
        {"type":"video", "videoID":"vid006"},
        {   "type":"video",
            "videoID":"vid007",
            "links": [
            {"type":"video", "videoID":"vid008", "links": [
                {   "type":"video",
                    "videoID":"vid009",
                    "links": [{"type":"video", "videoID":"vid010"}]
                }
            ]}
        ]},
    ]
}

def find_video(data):
    result = []
    if type(data) is dict:
        for key in data.keys():
            if key == 'videoID':
                ap = data.pop(key)
                result.append(ap) + find_video(data)
            if type(key) != str and key:
                data = key
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!





find_video(data)

videoID
vid001


ФИЛЬТРЫ, РАБОТА С СЛОВАРЕМ

In [ ]:
cheese_data = {
    'чеддер': [370, 5000, 33, 'твердый'],
    'пармезан': [510, 4000, 29, 'твердый'],
    'гауда': [250, 3700, 27, 'полутвердый'],
    'эдам': [220, 10000, 30, 'полутвердый'],
    'горгонзола': [320, 3000, 32, 'полумягкий'],
    'рокфор': [340, 15000, 31, 'полумягкий'],
    'стилтон': [360, 7000, 35, 'полумягкий'],
    'камамбер': [250, 8000, 24, 'мягкий'],
    'бри': [310, 6500, 28, 'мягкий'],
}
def filter_by_fat(cheese_data, n):
    filter_func = lambda x : cheese_data[x][2] < n
    key_ = list(filter(filter_func, cheese_data))
    dict_true = dict()
    for k in key_:
        dict_true.update({k: cheese_data[k][2]})
    return dict_true



print(filter_by_fat(cheese_data, n=30))
# {'пармезан': 29, 'гауда': 27, 'камамбер': 24, 'бри': 28}

#print(filter_by_fat(cheese_data, n=25))
# {'камамбер': 24}


{'чеддер': 33, 'горгонзола': 32, 'рокфор': 31, 'стилтон': 35}


In [ ]:
## Введите свое решение ниже
cheese_data = {
    'чеддер': [370, 5000, 33, 'твердый'],
    'пармезан': [510, 4000, 29, 'твердый'],
    'гауда': [250, 3700, 27, 'полутвердый'],
    'эдам': [220, 10000, 30, 'полутвердый'],
    'горгонзола': [320, 3000, 32, 'полумягкий'],
    'рокфор': [340, 15000, 31, 'полумягкий'],
    'стилтон': [360, 7000, 35, 'полумягкий'],
    'камамбер': [250, 8000, 24, 'мягкий'],
    'бри': [310, 6500, 28, 'мягкий'],
}
def count_money(cheese_data):
    map_func = lambda x: cheese_data[x][0]*10*cheese_data[x][1] # в списке res находятся произведения cheese_data[x][0]* cheese_data[x][1]
    res = sum(list(map(map_func, cheese_data)))
    return res
print(count_money(cheese_data))
## 196100000

19610000000


In [ ]:
## Введите свое решение ниже
cheese_data = {
    'чеддер': [370, 5000, 33, 'твердый'],
    'пармезан': [510, 4000, 29, 'твердый'],
    'гауда': [250, 3700, 27, 'полутвердый'],
    'эдам': [220, 10000, 30, 'полутвердый'],
    'горгонзола': [320, 3000, 32, 'полумягкий'],
    'рокфор': [340, 15000, 31, 'полумягкий'],
    'стилтон': [360, 7000, 35, 'полумягкий'],
    'камамбер': [250, 8000, 24, 'мягкий'],
    'бри': [310, 6500, 28, 'мягкий'],
}
def find_cheese_type(cheese_data, cheese_type):
    filter_func = lambda x: cheese_data[x][3] == cheese_type # пишем условие совпадения запроса с табличными данными
    map_func = lambda x: x # берем только название
    res = list(map(map_func, filter(filter_func, cheese_data) ))
    return res

find_cheese_type(cheese_data, cheese_type='мягкий')
# ['камамбер', 'бри']

find_cheese_type(cheese_data, cheese_type='полумягкий')
# ['горгонзола', 'рокфор', 'стилтон']


['горгонзола', 'рокфор', 'стилтон']

In [ ]:
cheese_data = {
    'чеддер': [370, 5000, 33, 'твердый'],
    'пармезан': [510, 4000, 29, 'твердый'],
    'гауда': [250, 3700, 27, 'полутвердый'],
    'эдам': [220, 10000, 30, 'полутвердый'],
    'горгонзола': [320, 3000, 32, 'полумягкий'],
    'рокфор': [340, 15000, 31, 'полумягкий'],
    'стилтон': [360, 7000, 35, 'полумягкий'],
    'камамбер': [250, 8000, 24, 'мягкий'],
    'бри': [310, 6500, 28, 'мягкий'],
}
def sort_cheese(cheese_data):
    map_func = lambda x: [x, cheese_data[x][0]]
    res_tuple = sorted(list(map(map_func, cheese_data)), key= lambda x: x[1])
    map_func_2 = lambda x: x[0]
    res = list(map(map_func_2, res_tuple))
    return res

print(sort_cheese(cheese_data))
# ['эдам', 'гауда', 'камамбер', 'бри', 'горгонзола', 'рокфор', 'стилтон', 'чеддер', 'пармезан']


['эдам', 'гауда', 'камамбер', 'бри', 'горгонзола', 'рокфор', 'стилтон', 'чеддер', 'пармезан']


ЗАДАЧА НА ПОИМКУ И ВЫВОД ДУБЛИКАТОВ
Если все ингредиенты разные, выводится строка "Ваш заказ оформлен верно".
Если ингредиент дублируется, выводится строка "Вы продублировали ингредиент {название ингредиента} в заказе {количество дублей} раз(а)"

In [ ]:
ingredients= ['кислота уксусная', 'кислота лимонная', 'закваска', 'кислота молочная', 'пряность', 'бактерии', 'аннато', 'кальций', 'калий', 'специя', 'молоко коровье', 'молоко овечье', 'фермент', 'соль', 'сливки', 'грибки', 'ароматизатор', 'молоко козье', 'дрожжи', 'каротин']

def purchase(ingredients):
    dict_count = dict(zip(ingredients, list(map(lambda x: 0, ingredients) ))) # создаем словарь. вторым списком будет список нулей
    list(map(lambda x: dict_count.update({x: dict_count[x]+1}), ingredients)) # заполняем словарь
    filter_func = lambda y: dict_count[y] > 1 # последовательно смотрим какие ингридиенты занесены несколько раз
    error_dict = list(filter(filter_func, dict_count))
    print("Ваш заказ оформлен верно") if not error_dict else print(*list(map(lambda p: f'Вы продублировали ингредиент {p} в заказе {dict_count[p]} раз(а)', error_dict)), sep= '\n')

purchase(ingredients)
# Ваш заказ оформлен верно


Ваш заказ оформлен верно
